django/http/multipartparser.py

```diff
@@ -166,15 +175,32 @@ def parse(self):
                 field_name = force_text(field_name, encoding, errors='replace')
 
                 if item_type == FIELD:
+                    # Avoid storing more than DATA_UPLOAD_MAX_NUMBER_FIELDS
+                    num_post_keys += 1
+                    if (settings.DATA_UPLOAD_MAX_NUMBER_FIELDS is not None and
+                            settings.DATA_UPLOAD_MAX_NUMBER_FIELDS < num_post_keys):
+                        raise TooManyFieldsSent('Too many fields sent. Check DATA_UPLOAD_MAX_NUMBER_FIELDS.')
+
+                    # Avoid reading more than DATA_UPLOAD_MAX_MEMORY_SIZE
+                    if settings.DATA_UPLOAD_MAX_MEMORY_SIZE is not None:
+                        read_size = settings.DATA_UPLOAD_MAX_MEMORY_SIZE - self._data_size
+
                     # This is a post field, we can just set it in the post
                     if transfer_encoding == 'base64':
-                        raw_data = field_stream.read()
+                        raw_data = field_stream.read(size=read_size)
                         try:
                             data = base64.b64decode(raw_data)
                         except _BASE64_DECODE_ERROR:
                             data = raw_data
                     else:
-                        data = field_stream.read()
+                        data = field_stream.read(size=read_size)
+
+                    # We add 2 here to make the check consistent with the GET
+                    # check that also includes '&='
+                    self._data_size += len(field_name) + len(data) + 2
```

Comment:
Unrelated to the above, but still on this line -- there appears to be some inconsistency in the treatment of base64-encoded values: `read_size` is applied to the raw (encoded) stream, but when the size is updated, it is according to the decoded data. Since the length difference between the two is significant (encoded is 33% longer), there's a subtle and hard-to-reproduce bug lurking here:
if data whose (decoded) length is around 75% of `DATA_UPLOAD_MAX_MEMORY_SIZE` is uploaded, the field will be cut (because the encoded length is more than the limit); if the cut length is right (chance of 1 in 12?) the decoding will succeed, and the field will be considered valid -- and some data simply lost.